File: Algorithm/Parametric Search/BOJ_2343.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

int solution(int N, int M, vector<int> bluelays) {
    int answer = 0;
    int l = *max_element(bluelays.begin(), bluelays.end()); 
    int r = accumulate(bluelays.begin(), bluelays.end(), 0);

    while (l <= r) {
        int mid = (l + r) / 2;
        int tot = 0, cnt = 1;

        for (int bluelay: bluelays) {
            if (bluelay > mid) {
                cnt = M + 1;
                break;
            }

            if (bluelay + tot > mid) {
                cnt++;
                tot = 0;
            }
            tot += bluelay;
        } // for

        if (cnt > M) {
            l = mid + 1;
        } else {
            r = mid - 1;
        }
    } // while

    answer = l;
    return answer;
}

int main(void) {
    int N, M;
    vector<int> bluelays;

    cin >> N >> M;
    bluelays = vector<int>(N, 0);

    for (int i = 0; i < N; ++i) {
        cin >> bluelays[i];
    }

    cout << solution(N, M, bluelays);
    return 0;
}

File: Algorithm/Parametric Search/BOJ_2343.swift
import Foundation

let NM = readLine()!.split(separator: " ").map { Int(String($0))! }
let bluelays = readLine()!.split(separator: " ").map { Int(String($0))! }

print(solution(NM[0], NM[1], bluelays))

func solution(_ N: Int, _ M: Int, _ bluelays: [Int]) -> Int {
  var answer = 0
  var l = bluelays.max()!, r = bluelays.reduce(0, +)

  while l <= r {
    let mid = (l + r) / 2
    var tot = 0, cnt = 1

    for bluelay in bluelays {
      if bluelay > mid {
        cnt = M + 1
        break
      }
      
      if tot + bluelay > mid {
        cnt += 1
        tot = 0
      }
      tot += bluelay
    } // for

    if (M < cnt) {
      l = mid + 1
    } else {
      r = mid - 1
    }
  } // while

  answer = l
  return answer
}